### ISSUE TITLE
[Docs] It's not clear (to me) from the docs how to deploy one's own auspice build

### ISSUE THREAD

**Issue by terrycojones** [nextstrain/auspice#1028]
Hi.  I'm trying to deploy auspice on a server of my own. I almost have it working. I can look at datasets and narratives, but I've not managed to get the top-level URL to work.

Should I be running `auspice build` locally (where I am putting the datasets and narrative together using augur), or should I transfer those directories to the server and run `auspice build` there? I assume the latter. Am I supposed to put the `dist` directory somewhere in particular or should it just be found?

At the moment I've done the auspice build on the server, but when I load the top-level URL in the browser Network tab I see an error

```
Request URL:https://myurl.org/charon/getDataset?prefix=/auspice/
Request Method:GET
Remote Address:18.194.225.45:443
Status Code:
500
```

In my shell where I'm running the server, I see output:

```
GET DATASET query received: prefix=/auspice/
Trace: Error: auspice not in available datasets
    at Object.extendDataPathsToMatchAvailable (/1GB/miniconda3/envs/auspice/lib/node_modules/auspice/cli/server/getDatasetHelpers.js:59:11)
    at /1GB/miniconda3/envs/auspice/lib/node_modules/auspice/cli/server/getDataset.js:9:15
    at /1GB/miniconda3/envs/auspice/lib/node_modules/auspice/cli/server/getDataset.js:13:15
[warning]       couldn't fetch JSONs -- auspice not in available datasets
```

I don't think there's anything wrong with the nginx rule I have in place to ship the request to the internal server, because a direct curl request to localhost gets a similar result:

```sh
$ curl -v 'http://localhost:4000/charon/getDataset?prefix=/auspice/'
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 4000 (#0)
> GET /charon/getDataset?prefix=/auspice/ HTTP/1.1
> Host: localhost:4000
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 500 couldn't fetch JSONs
< X-Powered-By: Express
< Date: Mon, 30 Mar 2020 12:39:33 GMT
< Connection: keep-alive
< Transfer-Encoding: chunked
< 
* Connection #0 to host localhost left intact
```

I guess there's some crucial config or step that I'm missing, but I don't know what....

Hmm, I just tried making a symlink called auspice.json in my datasets dir that points to one of the existing datasets, did an `auspice build` (on the server), and now I no longer get an error. So I guess an "auspice" dataset is looked for by default when a request comes in for the top-level?

Thanks for any help.  You've done a really great job putting all this together!

**Comment by terrycojones** [comment]
Although my workaround stops an error from happening, I no longer see a splash page that lists the available datasets and narratives, so I don't think I've fixed the problem at all - I've essentially just added another dataset.  Suggestions welcome......

**Comment by terrycojones** [comment]
In my original setup (i.e., without the fake 'auspice' dataset just mentioned), the sequence of events when I try to go to the top-level URL of my site is as follows

1. Client goes to http://myurl.com/auspice
1. nginx sends the request on to http://localhost:4000/  The response is a short index.html doc that contains `<script async src="/dist/auspice.bundle.js"></script>`
1. Next request is for /dist/auspice.bundle.js and that succeeds
1. There are some calls for e.g., `/dist/auspice.chunk.3.bundle.js` which also succeed
1. Then a request for `/charon/getDataset?prefix=/auspice/` which fails (500 error code, as mentioned above).

On localhost I can `curl http://localhost:4000/charon/getAvailable` and the response is JSON that contains the expected one dataset and one narrative. The dataset is not called `auspice` BTW.

So my question is: why, in the final list of requests above, is there an apparently hard-coded `?prefix=/auspice/`?  I wonder if that could be being passed along by nginx as part of its proxying (see step 1 above, where the external URL of my site ends in `/auspice`). Hmm... I'll experiment by changing the name and seeing if the `prefix` requested also changes. If so maybe I can tell nginx not to pass along that part of the original URL.

It would be very helpful for debugging if `auspice view --verbose` printed the actual HTTP conversation (headers, response). Then I could be sure about what was actually being requested and replay requests using curl to localhost to test if having nginx in the loop is creating the problem.

**Comment by terrycojones** [comment]
The answer to the above is that if I change the base external URL of the site (e.g, to `http://myurl.com/auspice6`) that the `getDataset` call that comes in specifies a correspondingly changed `prefix=/auspice6/`. So auspice is somehow getting that information from nginx.  I tried running with `proxy_pass_request_headers off;` and `proxy_hide_header Referer;` but that doesn't fix it.

It seems most likely to me (now) that the auspice client JS is generating the `?prefix=...` part of the request based on original request URL.  But why would it not ask for the list of datasets and instead just ask for a non-existent one using the basename of the original URL? Maybe there's an assumption in the client code that the server is always deployed at the top level of a domain and that any extra text in the URL must be the name of a dataset?

**Comment by terrycojones** [comment]
I think what I need is to use `--extend config.json` on the auspice build line, and to put a `serverAddress` in there, based on https://nextstrain.github.io/auspice/customise-client/api  However when I do that, I get webpack errors. E.g.,

```
WARNING in /1GB/ncov/dist.orig/auspice.chunk.9.bundle.js.map 1:10                                                                                          [23/1990]
Module parse failed: Unexpected token (1:10)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
> {"version":3,"file":"auspice.chunk.9.bundle.js","sources":["webpack:///auspice.chunk.9.bundle.js"],"mappings":"AAAA","sourceRoot":""}
 @ /1GB/ncov sync ^\.\/.*$ ./dist.orig/auspice.chunk.9.bundle.js.map
 @ ./src/util/extensions.js
 @ ./src/util/googleAnalytics.js
 @ ./src/index.js
 @ multi babel-polyfill ./src/index

ERROR in /1GB/ncov/dist.orig/auspice.chunk.3.bundle.js
Module not found: Error: Can't resolve '@babel/runtime/helpers/typeof' in '/1GB/ncov/dist.orig'
 @ /1GB/ncov/dist.orig/auspice.chunk.3.bundle.js 1:0-52 93:31-38 121:131-138 180:35-42 212:131-138 390:31-38 416:131-138
 @ /1GB/ncov sync ^\.\/.*$
 @ ./src/util/extensions.js
 @ ./src/util/googleAnalytics.js
 @ ./src/index.js
 @ multi babel-polyfill ./src/index
```

**Comment by terrycojones** [comment]
I would close this issue (unless someone wants it open?) as I've made my own landing page with links to the datasets & narratives I want to show.

**Comment by jameshadfield** [comment]
> assumption in the client code that the server is always deployed at the top level of a domain and that any extra text in the URL must be the name of a dataset?

This is the assumption for the default server provided with auspice. If you are using auspice in your own server, the idea is to [write your own handlers for the different GET requests](https://nextstrain.github.io/auspice/server/introduction#customising-the-default-auspice-server), which could respond appropriately to different URL structures, dataset naming formats etc. The `serverAddress` extension was designed to specify where these handlers were.

If you have improvements to the docs after having successfully done this they would be much appreciated 👍

**Comment by terrycojones** [comment]
Thanks @jameshadfield  One thing I discovered the hard way (well, not so hard, just digging into the code a bit) was that I needed to stick `env HOST=0.0.0.0` at the start of the command line so express would listen on all interfaces (this is what you do with your Heroku setup, in some params file which is maybe just specifying env vars). I looked at putting it into your docs, and it seems there's an obvious place but I wasn't sure if you'd want it there (the description on the page is quite minimal, and I thought it might be meant to be that way).

It took me a while to figure out exactly what was going wrong as I was doing this on an Amazon EC2 instance and was at first looking at the virtual network ACL, the Security Group, the Linux ufw, etc. In the end a friend suggested I run `tcpdump` on the EC2 instance interface, and I saw the packet arrive (and an RST response go back, triggering the connection denied), at which point it became clear....  So documenting this might save someone else some hours.

**Comment by jameshadfield** [comment]
> the description on the page is quite minimal, and I thought it might be meant to be that way

No, it's just because finding the time to write comprehensive documentation is hard! Please add it if you can  😄 

> So documenting this might save someone else some hours.

Yeah -- if you write a markdown tutorial of how you got it working on EC2 we'd love to add that as I'm sure other people would be interested. 

P.S. Did you ever get the splash page working? It may that it's hardcoded for a pathname of `/`, which should be customisable. A lot of auspice's design is understandable in the context of it being designed specifically for nextstrain.org, but over time I'd like to make it more general for use cases such as yours.

**Comment by terrycojones** [comment]
Yes, I agree (re hard) :-)

OK, I'll try to do that re EC2. It was pretty easy, thanks to your docs, apart from the HOST=0.0.0.0 thing (which of course only matters if you expose the auspice server to the outside world, not if you're sitting behind nginx and it's proxying to localhost).  So I could also mention deploying behind nginx (also very easy).

The splash page I worked around by just making my own, which is what I figured you'd done on nextstrain. I was going to make it do a `glob.glob` of `*.json` in the datasets dir, then send a request for each dataset and see which ones require BasicAuth (I have some that initially will not be public) and just list those that are not password protected (by nginx). But in the end I just hard-coded a list of the narratives and datasets I want to give access to.

I understand re building something then seeing how it should be refactored to be more general / portable / widely useful, etc. It's a long process and things only become clear gradually, and then can require a ton of time - all while balancing other stuff. Anyway, I find the logical division of the various parts very clean.

BTW, I made lots of suggested tiny changes to the `ncov` repo a couple of weeks ago while I drank a beer :-)  I had no idea how the code hung together at that point (still don't actually), but thought I could modernize some of the Python. So there are a bunch of prs over there. They could be closed, cherry-picked, whatever. I submitted one per file hoping that might make things more palatable :-) Many small changes were to make the code pass flake8 cleanly.
